### PR TITLE
DEV: Wiring for the admin about config page

### PR DESCRIPTION
**Pull request by OsamaSayegh**
This PR continues work laid out by https://github.com/discourse/discourse/pull/27340 for the admin config page for the /about page. The last PR set up the user interface, and this one sets up all the wiring needed to make the input fields and save buttons actually work.

Screenshot:

localhost_4200_admin_config_about

### CODE REVIEW

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/contact-information.gjs
Comment (ZogStriP) [inline app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/contact-information.gjs:72]: Should this be done in a `finally` in case anything bad happens during the `ajax` request?

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs
Comment (ZogStriP) [inline app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs:35]: Ditto ^^ Might be worth moving into a `finally`?

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/your-organization.gjs
Comment (ZogStriP) [inline app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/your-organization.gjs:46]: `finally`?

**Review by ZogStriP — APPROVED** [review]

File: app/assets/javascripts/admin/addon/routes/admin-config-about.js
Comment (CvX) [inline app/assets/javascripts/admin/addon/routes/admin-config-about.js:29]: This shouldn't be needed (the default implementation of setupController does just that)
```suggestion
```

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/your-organization.gjs
Comment (CvX) [inline app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/your-organization.gjs:30]: In all those cases you can use this in the template instead:

```hbs
{{on "input" (withEventValue (fn (mut this.companyName)))}}
```

File: app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs
Comment (CvX) [inline app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs:20]: This would be simpler:

```suggestion
  generalSettings = {
    title: this.#lookupSettingFromData("title"),
    siteDescription: this.#lookupSettingFromData("site_description"),
    extendedSiteDescription: this.#lookupSettingFromData(
      "extended_site_description"
    ),
    aboutBannerImage: this.#lookupSettingFromData("about_banner_image"),
  };
```

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/contact-information.gjs
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/contact-information.gjs:72]: Yes I agree with Regis here.

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs:35]: Yes I agree with Regis here.

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/contact-information.gjs
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/contact-information.gjs:165]: Rather than all these `showSavedAlert` checks + messages, we should utilize toasts instead:

```javascript
this.toasts.success({
  duration: 3000,
  data: { message: I18n.t("BLAH") },
});
```

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs:35]: Actually we won't need them if we use toasts instead

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/your-organization.gjs
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/your-organization.gjs:30]: So is this what should be used at all times now in Glimmer components? Should `companyName` and the others be @tracked or not? Is this the replacement for the "classic" 2 way data binding? It's a bit of a moot point with FormKit upcoming I guess but it would be good to have guidance on this.

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs:42]: What CVX said here https://github.com/discourse/discourse/pull/27492/files#r1646507402

File: config/locales/client.en.yml
Comment (martin-brennan) [inline config/locales/client.en.yml:5471]: We can remove this if we change to toasts

File: config/locales/server.en.yml
Comment (martin-brennan) [inline config/locales/server.en.yml:1705]: No need to mark these experimental; I think just make them `hidden: true` and only make them editable from the about config area for now.

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/contact-information.gjs
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/contact-information.gjs:163]: Let's make this more specific for config areas, should be `admin-config-area-card__btn-save`

File: app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs:63]:
```suggestion
          class="admin-config-area-about__general-settings-section"
```

File: app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs:73]:
```suggestion
          class="admin-config-area-about__contact-information-section"
```

File: app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs:83]:
```suggestion
          class="admin-config-area-about__your-organization-section"
```

File: app/controllers/admin/config/about_controller.rb
Comment (martin-brennan) [inline app/controllers/admin/config/about_controller.rb:16]: I think we should be using `PrettyText.cook` here?

File: app/controllers/admin/config/about_controller.rb
Comment (martin-brennan) [inline app/controllers/admin/config/about_controller.rb:10]: I think it would be easier on yourself if you made all the keys sent by the client the same as their actual site setting names, then you could do `settings_map.merge!(general_settings)` and so on.

File: app/controllers/admin/config/about_controller.rb
Comment (martin-brennan) [inline app/controllers/admin/config/about_controller.rb:39]: We need to be handling the `on_success` and `on_failed_policy` checks for this service:

https://github.com/discourse/discourse/blob/48c58499f1399baf88b43762d39bb0849eb00cf0/app/controllers/admin/site_settings_controller.rb#L45-L61

We also should write a request spec to make sure that doing this in a loop is okay and that failure is handled correctly.

File: app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs:49]: I think you could do `this.args.data.findBy("setting", name);` here

File: app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs:20]: Agree with this too for all 3 hashes.

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs:91]: Let's make this more specific for config areas, should be `admin-config-area-card__btn-save`

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/your-organization.gjs
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/your-organization.gjs:107]: Let's make this more specific for config areas, should be `admin-config-area-card__btn-save`

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs
Comment (martin-brennan) [reply app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/general-settings.gjs:91]: Also we should disable all save buttons while the AJAX call is happening

File: app/controllers/admin/config/about_controller.rb
Comment (martin-brennan) [inline app/controllers/admin/config/about_controller.rb:43]: When we hide the other settings (the ones marked EXPERIMENTAL) this will need to be more like `[:setting1, :setting2, :setting3].includes?(name)`

**Review by martin-brennan — CHANGES_REQUESTED** [review]
Heading in the right direction, form looks nice but just needs a few changes.

FYI @jjaffeux is likely going to convert these forms later to use FormKit once it lands to iron out more use cases. It's fine to merge this as is though (after fixups), it's just so you are aware.

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/your-organization.gjs
Comment (CvX) [reply app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/your-organization.gjs:30]: Yes, two-way bindings are out. So you should use either `withEventValue` helper or define actions (if you need to transform data in some way)

re: `@tracked` - it depends if you use those values to render anything, directly or indirectly - through other properties that read them. (that's what this decorator pretty much does - "please update the UI whenever this changes")

I don't know yet how exactly it will look like with form-kit (still have to review it 😅) but I'd say those apply generally in most parts of the app.

File: app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/contact-information.gjs
Comment (OsamaSayegh) [reply app/assets/javascripts/admin/addon/components/admin-config-area-cards/about/contact-information.gjs:72]: No, the "saved!" indicator should only show up when the ajax calls succeeds so we don't want to set `showSavedAlert` in a `finally` block.

File: app/controllers/admin/config/about_controller.rb
Comment (OsamaSayegh) [reply app/controllers/admin/config/about_controller.rb:10]: There are a couple of reasons why I went with the current approach:

1. I didn't want to make it possible to change arbitrary site settings via this API endpoint. If I merged the `params` object directly into the `settings_map`, then it would be possible for a client to shove in an unrelated site setting and the change would go through. It wouldn't be a security issue, but feels unhygienic.

2. Some setting names don't match their field names in the UI because we're reusing existing settings, but I think it will be nice to have the params for the API endpoint match the names of the fields in the UI.

File: config/locales/client.en.yml
Comment (OsamaSayegh) [reply config/locales/client.en.yml:5471]: I need to provide a message for the toast, what should the toast say?

File: app/controllers/admin/config/about_controller.rb
Comment (OsamaSayegh) [reply app/controllers/admin/config/about_controller.rb:39]: I don't know what makes sense to do inside `on_success` here? Calling `SiteSettingUpdateExistingUsers` here would essentially be a no-op because it only applies to specific site settings which we don't touch here.

It's also not possible for the `setting_is_visible` or `setting_is_configurable` policies to fail here because we change specific settings here none of which are hidden or unconfigurable (the ones that are we will `allow_changing_hidden: true` for them).

File: app/controllers/admin/config/about_controller.rb
Comment (OsamaSayegh) [reply app/controllers/admin/config/about_controller.rb:16]: I looked at both methods and saw that `.cook` calls `.markdown` and then adds mentions and video support so I went with `markdown`. Should mention/videos be supported in the extended description?

Comment (martin-brennan) [reply app/controllers/admin/config/about_controller.rb:16]: Good point, you are right that the simpler one is better :+1:

File: config/locales/server.en.yml
Comment (OsamaSayegh) [reply config/locales/server.en.yml]: I made the new settings hidden, but now I've realized that hidden settings cannot be retrieved via the `Admin::SiteSettingsController#index` endpoint which is called by the about config page and specs are failing because of that.

I'm not super sure about adding `include_hidden: true` here:

https://github.com/discourse/discourse/blob/4a6b79dead474c459875d6342b7ffd90b462716d/app/controllers/admin/site_settings_controller.rb#L14-L19

because there are probably some settings that are super secret that even admins should be able to see. What we should we do here?

File: app/controllers/admin/config/about_controller.rb
Comment (martin-brennan) [reply app/controllers/admin/config/about_controller.rb:39]: I think just move `render json: success_json` into `on_success`, then add a general `on_failure` call like this chat endpoint to handle anything else:

https://github.com/discourse/discourse/blob/e92a82aa1d16e24a9447930d0d3441716b40efaa/plugins/chat/app/controllers/chat/api/channels_controller.rb#L39-L40

Main point is we need to handle the `on_X` stuff explicitly

File: app/controllers/admin/config/about_controller.rb
Comment (martin-brennan) [reply app/controllers/admin/config/about_controller.rb:10]: Okay yes I see your reasoning now. Let's leave it as is for now then, maybe later on as we do more of these mixed setting config pages we can think of some sensible pattern to take bulk keys/values from the UI and convert them to a big hash of only allowed settings to update for a certain endpoint :+1:

File: config/site_settings.yml
Comment (martin-brennan) [inline config/site_settings.yml:417]: I think since these are hidden now you need to add `include_hidden: true` at `Admin::SiteSettingsController#index`

Comment (martin-brennan) [reply config/site_settings.yml:417]: Otherwise the page is currently broken

File: config/locales/client.en.yml
Comment (martin-brennan) [reply config/locales/client.en.yml:5471]: Can we do it based on the section? Eg. "General settings saved", "Contact information saved" etc.? If that's too annoying just do "About configuration saved"

File: config/locales/server.en.yml
Comment (martin-brennan) [reply config/locales/server.en.yml]: Ah I didn't see this comment, and I just sent a comment about `include_hidden`. Hmm it's quite tricky, and you are right we probably shouldn't make it too broad.

What do you think about this? We add a `ADMIN_CONFIG_AREA_ALLOWLISTED_HIDDEN_SETTINGS` const in `Admin::SiteSettingsController`, and then add another param to `all_settings` called `filter_allowed_hidden` that passes that array in? Then if `include_hidden` is `true` and `filter_allowed_hidden` is passed in, we filter out any hidden settings that aren't in `filter_allowed_hidden`

Kind of weird and a bit messy but I don't see a better way for now. Also...for these particular settings it's only temporary, once we finish all this about page work we can make these site settings visible.

File: config/site_settings.yml
Comment (martin-brennan) [reply config/site_settings.yml:417]: NVM this I saw + responded to your other comment about include_hidden :)

File: config/locales/server.en.yml
Comment (OsamaSayegh) [reply config/locales/server.en.yml]: I've done this in https://github.com/discourse/discourse/pull/27492/commits/45730d0039c77fb7c76592cded852618b28fcdc4.

File: app/controllers/admin/config/about_controller.rb
Comment (OsamaSayegh) [reply app/controllers/admin/config/about_controller.rb:39]: Hmm I just realized that because I'm currently calling the service for each setting inside a loop, I can't do any `render`ing within service callbacks because it's not allowed (nor does it make sense) to `render` multiple times in an action. Tricky stuff...

File: app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs
Comment (martin-brennan) [inline app/assets/javascripts/admin/addon/components/admin-config-areas/about.gjs:43]: This is interesting, I think this is a good way of handling it, maybe at some point it makes sense to add some generic service for this so you could do something like `this.configAreaService.setSaving("about")` then do `this.configAreaService.getSavingStatus("about")` in child components. Anyway that's a thought for the future

File: app/controllers/admin/config/about_controller.rb
Comment (martin-brennan) [reply app/controllers/admin/config/about_controller.rb:39]: Ah right that is a pickle :thinking: I think really the ultimate fix would be to make a service that updates multiple site settings to new values. This way you could run the policies once for all settings, and the success/failure would only be once too since you are updating all the settings (and adding logs) in a single transaction.

For the sake of getting this PR over the line, since it's still hidden behind various settings, let's punt on this for now. I would like you to start thinking of this service though. You will likely need this other feature of services, where you can give a separate class to a policy:

https://github.com/discourse/discourse/blob/0b3b101c97b6a027ee11bcd166d1ab49a63a0c00/plugins/chat/app/services/chat/create_direct_message_channel.rb#L29-L30

https://github.com/discourse/discourse/blob/0b3b101c97b6a027ee11bcd166d1ab49a63a0c00/plugins/chat/app/policies/chat/direct_message_channel/max_users_excess_policy.rb

**Review by martin-brennan — APPROVED** [review]
Thanks for bearing with me Osama, let's get this merged so you can move on with other parts of this.

**Comment by OsamaSayegh** [comment]
thanks for the through reviews @martin-brennan, merging this is in now 🚀

**Pull request merged**
